python/ray/data/_internal/stats.py

```diff
@@ -201,6 +210,8 @@ def __init__(
         self.time_total_s: float = 0
         self.needs_stats_actor = needs_stats_actor
         self.stats_uuid = stats_uuid
+        # Dict of summary statistics aggregated by stage. Keyed by stage_name
+        self.stage_stats: Dict[str, Dict[str, Any]] = {}
```

Comment:
Instead of caching the stats for all stages on the `DatasetStats` object, could we turn `DatasetStats._summarize_blocks()` into a static method or utility function in this module that takes in the dict of stats for a stage? I.e. instead of
```python
self._calculate_blocks_stats(stage_name, metadata, is_substage=False)
blocks_stats_str = self._summarize_blocks(stage_name, is_substage=False)
```
it would be
```python
block_stats = self._calculate_blocks_stats(stage_name, metadata, is_substage=False)
blocks_stats_str = DatasetStats._summarize_blocks(block_stats, is_substage=False)
```

There are a few pros to this approach:
1. No new state added to the `DatasetStats` object, and we avoid adding new state mutation to the class; this keeps the `DatasetStats` object's data model simple.
2. Lifecycle of the stage stats dict is tied directly to its use, instead of being tied to the `DatasetStats` object (which will usually live a lot longer than the stats dict's use).
3. The `_summarize_blocks` contract is more explicit (block stats dict --> summary string vs. stage name --> internally look up block stats dict --> summary string); this should also make this easier to unit test, if so desired.

The only con that I can think of is the calculation of the stats dict object not being cached on repeated calls, but I don't think that's a concern (shouldn't be all that expensive, and repeated calls shouldn't be typical).

Reply:
@c21 did we want to cache the calculated stats as a class attribute here to use in other methods? or are we OK with calculating the stats on the fly each time?

Reply:
@clarkzinzow - I agree with the overall code structure. @scottjlee - could you help make the code change accordingly?

> The only con that I can think of is the calculation of the stats dict object not being cached on repeated calls, but I don't think that's a concern (shouldn't be all that expensive, and repeated calls shouldn't be typical).

This is not a concern now, but will be one issue later when we have larger scale (other system as Spark always caches the stage's aggregated metadata). But I agree this looks like a preoptimization, we can defer this to later when needed (especially we will move to Core metrics integration).